fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/PrepareCollectMetaTask.java

```diff
@@ -0,0 +1,87 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.optimizer.task;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.starrocks.common.profile.Timer;
+import com.starrocks.common.profile.Tracers;
+import com.starrocks.connector.WarmupEntry;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.server.MetadataMgr;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalScanOperator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import static com.starrocks.common.profile.Tracers.Module.EXTERNAL;
+
+public class PrepareCollectMetaTask extends OptimizerTask {
+
+    private final OptExpression planTree;
+
+    public PrepareCollectMetaTask(TaskContext context, OptExpression root) {
+        super(context);
+        this.planTree = root;
+        Preconditions.checkState(planTree.getOp().getOpType() == OperatorType.LOGICAL);
+    }
+
+    @Override
+    public void execute() {
+        List<LogicalScanOperator> scanOperators = collectScanOperators(planTree)
+                .stream()
+                .filter(scanOperator -> scanOperator.getTable().supportWarmupMetadata())
+                .collect(Collectors.groupingBy(LogicalScanOperator::getOpType))
+                .values()
+                .stream()
+                .filter(list -> list.size() > 1)
+                .flatMap(List::stream)
+                .collect(Collectors.toList());
+
+        if (scanOperators.isEmpty()) {
+            return;
+        }
+
+        int threadPoolSize = context.getOptimizerContext().getSessionVariable().getWarmupMetadataPoolSize();
+        String queryId = context.getOptimizerContext().getQueryId().toString();
+        ExecutorService executorService = Executors.newFixedThreadPool(threadPoolSize,
+                new ThreadFactoryBuilder().setNameFormat(String.format("warmup-metadata-%s", queryId)).build());
+
+        MetadataMgr metadataMgr = GlobalStateMgr.getCurrentState().getMetadataMgr();
+
+        try (Timer ignored = Tracers.watchScope(EXTERNAL, "EXTERNAL.parallel_warmup_metadata")) {
+            CompletableFuture<Void> allFutures = CompletableFuture.allOf(scanOperators.stream()
+                    .map(op -> CompletableFuture.supplyAsync(() ->
+                                    metadataMgr.warmupMetadata(queryId, op.getTable().getCatalogName(),
```

Comment:
what happens If one table's warmup metadata failed?

Reply:
It will return true immediately and  reenter the normal flow and throw an exception.